File: src/main/java/failchat/goodgame/GGChatClient.java
package failchat.goodgame;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import failchat.core.*;
import failchat.handlers.CommonHighlightHandler;
import failchat.handlers.MessageObjectCleaner;
import org.java_websocket.handshake.ServerHandshake;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class GGChatClient implements ChatClient {

    private static final Logger logger = Logger.getLogger(GGChatClient.class.getName());
    private static final String GG_WS_URL = "ws://chat.goodgame.ru:8081/chat/websocket";

    private WSClient wsClient;
    private ChatClientStatus status;
    private MessageManager messageManager = MessageManager.getInstance();
    private Moderation moderation = Moderation.getInstance();
    private List<MessageHandler<GGMessage>> messageHandlers;
    private String channelName;
    private int channelId;
    private ObjectMapper objectMapper;

    public GGChatClient(String channelName) {
        this.channelName = channelName;
        objectMapper = new ObjectMapper();
        messageHandlers = new ArrayList<>();
        //noinspection unchecked
        messageHandlers.add(MessageObjectCleaner.getInstance());
        messageHandlers.add(new UrlCleaner());
        messageHandlers.add(new GGSmileHandler());
        messageHandlers.add(new CommonHighlightHandler(channelName));
        status = ChatClientStatus.READY;
    }

    @Override
    public void goOnline() {
        if (status != ChatClientStatus.READY) {
            return;
        }
        channelId = GGApiWorker.getChannelIdByName(channelName);
        if (channelId == -1) {
            status = ChatClientStatus.ERROR;
            return;
        }
        wsClient = new GGWSClient();
        wsClient.connect();
    }

    @Override
    public void goOffline() {
        status = ChatClientStatus.SHUTDOWN;
        if (wsClient != null) {
            wsClient.close();
        }
    }

    @Override
    public ChatClientStatus getStatus() {
        return status;
    }

    private class GGWSClient extends WSClient {
        GGWSClient() {
            super(GG_WS_URL);
        }

        @Override
        public void onOpen(ServerHandshake serverHandshake) {
            try {
                String connectToChannelMes = objectMapper.writeValueAsString(new JoinWSMessage((channelId)));
                wsClient.send(connectToChannelMes);
                status = ChatClientStatus.WORKING;
                logger.info("Connected to goodgame");
                messageManager.sendInfoMessage(new InfoMessage(Source.GOODGAME, "connected"));
            } catch (JsonProcessingException e) {
                logger.log(Level.WARNING, "Something goes wrong...", e);
            }
        }

        @Override
        public void onMessage(String s) {
            try {
                GoodgameCommonMessage ggwsm = objectMapper.readValue(s, GoodgameCommonMessage.class);
                switch (ggwsm.getType()) {
                    case "message": {
                        handleUserMessage(ggwsm.getData());
                        break;
                    }
                    case "remove_message": {
                        handleModMessage(ggwsm.getData());
                        break;
                    }
                }
            } catch (IOException e) {
                logger.log(Level.WARNING, "Something goes wrong...", e);
            }
        }

        @Override
        public void onClose(int i, String s, boolean b) {
            logger.info("Goodgame disconnected");
        }

        @Override
        public void onError(Exception e) {
            logger.log(Level.WARNING, "Disconnect? ...", e);
        }

        @Override
        public void onReconnect() {
            logger.info("Goodgame disconnected, trying to reconnect ...");
            messageManager.sendInfoMessage(new InfoMessage(Source.GOODGAME, "disconnected"));
        }

        private void handleUserMessage(JsonNode messageObj) throws IOException {
            GGMessage message = objectMapper.convertValue(messageObj, GGMessage.class);
            for (MessageHandler<GGMessage> messageHandler : messageHandlers) {
                messageHandler.handleMessage(message);
            }
            messageManager.sendMessage(message);
        }

        private void handleModMessage(JsonNode messageObj) throws IOException {
            GoodgameDeleteMessage ggDelMes = objectMapper.convertValue(messageObj, GoodgameDeleteMessage.class);
            moderation.deleteGgMessage(ggDelMes.getMessageId());
        }
    }


    /**
     * Класс для сериализации в json. Используется для подключения к каналу.
     */
    private static class JoinWSMessage {
        private String type = "join";
        private JoinWSData data;

        public String getType() {
            return type;
        }

        public JoinWSData getData() {
            return data;
        }

        JoinWSMessage(int channelId) {
            data = new JoinWSData(channelId);
        }
    }

    /**
     * Класс для сериализации в json. Используется для подключения к каналу.
     */
    private static class JoinWSData {
        private String channelId;

        private boolean hidden = false;

        JoinWSData(int channelId) {
            this.channelId = Integer.toString(channelId);
        }

        @JsonProperty("channel_id")
        public String getChannelId() {
            return channelId;
        }

        public boolean isHidden() {
            return hidden;
        }
    }

    /**
     * Класс для десериализации из json'a. Представляет общий вид сообщения
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class GoodgameCommonMessage {
        private String type;
        private JsonNode data;

        public JsonNode getData() {
            return data;
        }

        public void setData(JsonNode data) {
            this.data = data;
        }

        public String getType() {
            return type;
        }

        public void setType(String type) {
            this.type = type;
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class GoodgameDeleteMessage {
        protected int messageId;

        public int getMessageId() {
            return messageId;
        }

        @JsonProperty("message_id")
        public void setMessageId(int messageId) {
            this.messageId = messageId;
        }
    }
}

File: src/main/kotlin/failchat/goodgame/GgChatClient.kt
package failchat.goodgame

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import failchat.Origin
import failchat.Origin.GOODGAME
import failchat.chat.ChatClient
import failchat.chat.ChatClientCallbacks
import failchat.chat.ChatClientStatus
import failchat.chat.ChatClientStatus.OFFLINE
import failchat.chat.ChatMessageHistory
import failchat.chat.MessageHandler
import failchat.chat.MessageIdGenerator
import failchat.chat.OriginStatus.CONNECTED
import failchat.chat.OriginStatus.DISCONNECTED
import failchat.chat.StatusUpdate
import failchat.chat.findFirstTyped
import failchat.chat.handlers.CommaHighlightHandler
import failchat.chat.handlers.ElementLabelEscaper
import failchat.ws.client.WsClient
import kotlinx.coroutines.runBlocking
import mu.KotlinLogging
import org.java_websocket.handshake.ServerHandshake
import java.net.URI
import java.util.concurrent.atomic.AtomicReference

class GgChatClient(
        private val channel: GgChannel,
        private val webSocketUri: String,
        private val messageIdGenerator: MessageIdGenerator,
        emoticonHandler: MessageHandler<GgMessage>,
        badgeHandler: GgBadgeHandler,
        private val history: ChatMessageHistory,
        override val callbacks: ChatClientCallbacks,
        private val objectMapper: ObjectMapper
) : ChatClient {

    private companion object {
        val logger = KotlinLogging.logger {}
    }

    override val origin = Origin.GOODGAME
    override val status: ChatClientStatus get() = atomicStatus.get()


    private var wsClient: WsClient = GgWsClient(URI.create(webSocketUri))
    private val atomicStatus: AtomicReference<ChatClientStatus> = AtomicReference(ChatClientStatus.READY)

    private val messageHandlers: List<MessageHandler<GgMessage>> = listOf(
            ElementLabelEscaper(),
            HtmlUrlCleaner(),
            emoticonHandler,
            CommaHighlightHandler(channel.name),
            badgeHandler,
            GgAuthorColorHandler()
    )


    override fun start() {
        //todo change to connecting
        if (atomicStatus.get() != ChatClientStatus.READY) {
            return
        }
        wsClient.start()
    }

    override fun stop() {
        atomicStatus.set(OFFLINE)
        wsClient.stop()
    }

    private inner class GgWsClient(uri: URI) : WsClient(uri) {

        override fun onOpen(serverHandshake: ServerHandshake) {
            val joinMessage = objectMapper.createObjectNode().apply {
                put("type", "join")
                putObject("data").apply {
                    put("channel_id", channel.id.toString())
                    put("isHidden", false)
                }
            }
            atomicStatus.set(ChatClientStatus.CONNECTED)
            logger.info("Goodgame chat client connected to channel {}", channel.id)

            wsClient.send(joinMessage.toString())
            callbacks.onStatusUpdate(StatusUpdate(GOODGAME, CONNECTED))
        }

        override fun onClose(code: Int, reason: String, remote: Boolean) {
            logger.info("Goodgame chat client disconnected from channel {}", channel.id)
        }

        override fun onMessage(message: String) {
            val messageNode = objectMapper.readTree(message)
            val type = messageNode.get("type").asText()
            val data = messageNode.get("data")

            //todo log on unknown type
            when (type) {
                "message" -> handleUserMessage(data)
                "remove_message" -> handleModMessage(data)
            }
        }

        override fun onError(e: Exception) {
            logger.error("Goodgame chat client error", e)
        }

        override fun onReconnect() {
            logger.info("Goodgame chat client disconnected, trying to reconnect")
            callbacks.onStatusUpdate(StatusUpdate(GOODGAME, DISCONNECTED))
        }

        private fun handleUserMessage(dataNode: JsonNode) {
            val subscriptionDuration = dataNode.get("resubs").fields().asSequence()
                    .map { (channelId, duration) -> channelId.toLong() to duration.intValue() }
                    .toMap(HashMap())
            val ggMessage = GgMessage(
                    id = messageIdGenerator.generate(),
                    ggId = dataNode.get("message_id").asText().toLong(),
                    author = dataNode.get("user_name").asText(),
                    text = dataNode.get("text").asText(),
                    authorHasPremium = dataNode.get("premium").asBoolean(),
                    subscriptionDuration = subscriptionDuration,
                    badgeName = dataNode.get("icon").textValue(),
                    authorColorName = dataNode.get("color").textValue(),
                    sponsorLevel = dataNode.get("payments").intValue(),
                    authorRights = dataNode.get("user_rights").intValue()
            )

            messageHandlers.forEach { it.handleMessage(ggMessage) }

            callbacks.onChatMessage(ggMessage)
        }

        private fun handleModMessage(dataNode: JsonNode) {
            val idToRemove = dataNode.get("message_id").asText().toLong()

            val foundMessage = runBlocking {
                history.findFirstTyped<GgMessage> { it.ggId == idToRemove }
            }
            foundMessage?.let { callbacks.onChatMessageDeleted(it) }
        }
    }

}
